### PR TITLE
Unstoppabledomains/resolution

### PR DESCRIPTION
**Pull request by JohnnyJumper**
This PR adds Unstoppable domains integration into the flow of adding new safe addresses. 

Key points:
* Updates the UI from ENS only to the general "blockchain domain" 
* Uses previous ENS integration in the case when a domain is not ending on .crypto or .zil
* Uses UnstoppableDomainResolution to fetch the eth address on a stored domain.

### CODE REVIEW

**Comment by DmitryBespalov** [comment]
Hi @JohnnyJumper thanks for opening a pr!

File: Multisig/Logic/Ethereum/BlockchainDomainManager.swift
Comment (DmitryBespalov) [inline Multisig/Logic/Ethereum/BlockchainDomainManager.swift]: is this guaranteed not to throw?

File: Multisig/Logic/Ethereum/BlockchainDomainManager.swift
Comment (DmitryBespalov) [inline Multisig/Logic/Ethereum/BlockchainDomainManager.swift]: the indentation seem to be a bit off

File: Multisig/Logic/Ethereum/BlockchainDomainManager.swift
Comment (DmitryBespalov) [inline Multisig/Logic/Ethereum/BlockchainDomainManager.swift]: we could use `Address(exactly: address)` here to remove force unwrap

File: Multisig/Logic/Ethereum/BlockchainDomainManager.swift
Comment (DmitryBespalov) [inline Multisig/Logic/Ethereum/BlockchainDomainManager.swift]: I think having 2 if-lets would be more readable here:

```swift
if let error = err as? ResolutionError {
  throw self.throwCorrectUdError(error, domain)
} else if let error = err {
  throw error
}
```

File: Multisig/Logic/Ethereum/BlockchainDomainManager.swift
Comment (DmitryBespalov) [inline Multisig/Logic/Ethereum/BlockchainDomainManager.swift]: Let's rename `err` to `resolutionError`

File: Multisig/Logic/Ethereum/BlockchainDomainManager.swift
Comment (DmitryBespalov) [reply Multisig/Logic/Ethereum/BlockchainDomainManager.swift:45]: however, if the `address` is not guaranteed to be an address String, there's an option to use `try Address(from: address)`

File: Multisig/Logic/Ethereum/BlockchainDomainManager.swift
Comment (DmitryBespalov) [inline Multisig/Logic/Ethereum/BlockchainDomainManager.swift]: Please rename `err` to `error` to use a full name

File: Multisig/Logic/Ethereum/BlockchainDomainManager.swift
Comment (DmitryBespalov) [reply Multisig/Logic/Ethereum/BlockchainDomainManager.swift]: Alternative is to use `try Address(from: address)` if there's a possibility that the address string is not an Ethereum address

File: Multisig/Logic/Ethereum/BlockchainDomainManager.swift
Comment (JohnnyJumper) [reply Multisig/Logic/Ethereum/BlockchainDomainManager.swift]: @DmitryBespalov, As long as we are using default settings this won't throw, this, however, does limit the app to the mainnet only. 
I can reconfigure the library to use testnet as well for future development. We support Rinkeby for .crypto. 

Where can I get network configurations such as chainId and blockchain provider (I assume it is going to be infura as in gnosis-android) to do so?

Comment (DmitryBespalov) [reply Multisig/Logic/Ethereum/BlockchainDomainManager.swift]: Got it, yes, that would be good to have because we have versions of the app working on Rinkeby network (in the future also on other chains). 

To get the blockchain provider url: `App.configuration.services.ethereumServiceURL`

To get the network name (we don't have chainId specified) - `App.configuration.app.network`

Comment (JohnnyJumper) [reply Multisig/Logic/Ethereum/BlockchainDomainManager.swift]: Awesome, I have connected the library to the potential testnets and watched the error in case if network supplied is not supported by UD.

**Comment by DmitryBespalov** [comment]
@JohnnyJumper ok, so far it looks good to me. I'll ping other devs for review. 

One other thing though is that we're about to transition from CocoaPods to SwiftPM, does UD support it? (I see no Package.swift there).

**Comment by JohnnyJumper** [comment]
> @JohnnyJumper ok, so far it looks good to me. I'll ping other devs for review.
> 
> One other thing though is that we're about to transition from CocoaPods to SwiftPM, does UD support it? (I see no Package.swift there).

Should be relatively easy to update by adding this: 
```swift
package.dependencies.append(
    .package(url: "https://github.com/unstoppabledomains/resolution-swift", from: "0.3.5")
)
```
When I launched the project via xcode the first time I didn't notice any other packages in SwiftPM and went with cocoapods.

**Comment by DmitryBespalov** [comment]
Great! Yes, it's in progress in the current sprint (I'll work on it next) https://github.com/gnosis/safe-ios/issues/1041

File: Multisig/Logic/Ethereum/BlockchainDomainManager.swift
Comment (sche) [inline Multisig/Logic/Ethereum/BlockchainDomainManager.swift]: We don't use ';'. Could you please remove it everywhere in this file?

File: Multisig/Logic/Ethereum/BlockchainDomainManager.swift
Comment (sche) [inline Multisig/Logic/Ethereum/BlockchainDomainManager.swift]: guard let resolution = resolution else { ... }
No need to use force unwrapping below.

**Review by sche — CHANGES_REQUESTED** [review]
Lgtm after fixing minor things.
The major question here is alignment with Android app. Here the proposal is to use "Enter Blockchain domain", but on Android it seems a split of ENS/UD when selecting how to enter address.
We will clarify with product how to keep it and one of the platform realisations would need to change it.

**Comment by JohnnyJumper** [comment]
> Lgtm after fixing minor things.
> The major question here is alignment with Android app. Here the proposal is to use "Enter Blockchain domain", but on Android it seems a split of ENS/UD when selecting how to enter address.
> We will clarify with product how to keep it and one of the platform realisations would need to change it.

That is true. It slipped from my mind that both apps should have the same implementation. Let me know which one should I fix.

**Comment by sche** [comment]
@tschubotz what option would you prefer?

**Comment by tschubotz** [comment]
> @tschubotz what option would you prefer?

I'd prefer the Android way, i.e. keeping both separate. That way everything is just more explicit and it names the supported type (ENS, UD) instead of "Blockchain domains" which could mean many things :)

**Comment by JohnnyJumper** [comment]
@sche  @tschubotz I split the ens and ud into separate views, also updated the library to the latest version, and changed a few error messages to make them more explicit. Can you take a look?

**Comment by DmitryBespalov** [comment]
@JohnnyJumper we've also moved from CocoaPods to SwiftPM, can you please update this PR accordingly?

File: Multisig/UI/Safe Management/Add Safe/EnterENSNameViewController.swift
Comment (DmitryBespalov) [inline Multisig/UI/Safe Management/Add Safe/EnterENSNameViewController.swift]: I would keep the wording as it was before

File: Multisig/UI/Safe Management/Add Safe/EnterSafeAddressViewController/EnterSafeAddressViewController.swift
Comment (DmitryBespalov) [inline Multisig/UI/Safe Management/Add Safe/EnterSafeAddressViewController/EnterSafeAddressViewController.swift]: here too, I'd keep the existing wording intact

File: Multisig/UI/Safe Management/Add Safe/EnterENSNameViewController.swift
Comment (JohnnyJumper) [reply Multisig/UI/Safe Management/Add Safe/EnterENSNameViewController.swift]: Done

**Comment by JohnnyJumper** [comment]
@DmitryBespalov 
Merged the main branch into here, removed all cocoapods, and added ud library to SwiftPM
Also updated the text from domain to name in both ens and ud version of UI

File: Multisig/Data/Services/Utils/GSError.swift
Comment (DmitryBespalov) [inline Multisig/Data/Services/Utils/GSError.swift]: sorry, one more thing that I overlooked - let's revert this struct (and `howToFix` below, too)

**Review by DmitryBespalov — APPROVED** [review]
one minor fix left, otherwise looks good!

File: Multisig/Data/Services/Utils/GSError.swift
Comment (JohnnyJumper) [reply Multisig/Data/Services/Utils/GSError.swift]: What a great catch! I completely forgot to rename it back. thank you!

File: Multisig.xcodeproj/project.pbxproj
Comment (sche) [inline Multisig.xcodeproj/project.pbxproj]: (null) looks like smth broken

File: Multisig.xcworkspace/xcshareddata/swiftpm/Package.resolved
Comment (sche) [inline Multisig.xcworkspace/xcshareddata/swiftpm/Package.resolved]: why was it downgraded? Could you please update in your package `.package(url: "https://github.com/krzyzanowskim/CryptoSwift.git", from: "1.0.0")` to 1.4.0?

Comment (sche) [reply Multisig.xcworkspace/xcshareddata/swiftpm/Package.resolved]: probably it will be properly resolved without the update too

File: Multisig/Data/Services/Utils/GSError.swift
Comment (sche) [inline Multisig/Data/Services/Utils/GSError.swift]: Error codes should be aligned between iOS and Android. I did not check, are they?

File: Multisig/UI/Safe Management/Add Safe/EnterUnstoppableNameViewController.swift
Comment (sche) [inline Multisig/UI/Safe Management/Add Safe/EnterUnstoppableNameViewController.swift]: We need to add `.safeAddUD` event. And also please assure that it is the same on Android.

**Comment by sche** [comment]
Looks good! 👍🏻 
Please resolve conflicts and align error messages & tracking event for both iOS and Android and we are good to go.

File: Multisig.xcworkspace/xcshareddata/swiftpm/Package.resolved
Comment (JohnnyJumper) [reply Multisig.xcworkspace/xcshareddata/swiftpm/Package.resolved]: At the moment we can't update the CryptoSwift package.

```Because no versions of EthereumAddress match the requirement 1.3.1..<2.0.0 and EthereumAddress 1.3.0 depends 
on CryptoSwift 1.0.0..<1.1.0, EthereumAddress >=1.3.0 requires CryptoSwift 1.0.0..<1.1.0.
And because root depends on CryptoSwift 1.4.0..<2.0.0 and root depends on EthereumAddress 1.3.0..<2.0.0, 
version solving failed.
```

As I understood this doesn't cause an issue for the gnosis project.  We will work on the update soon

**Comment by JohnnyJumper** [comment]
@sche  There is something strange happening when I resolve the conflicts. Solving the conflicts triggered codecov report above saying that the coverage dropped. I believe I made some cruel mistakes in merging as the files are mostly xcode generated. Can you help me out here? 

P.S once I reverted the merge commit codecov edited its report saying that now the coverage drop is 0.40 instead of 63 something %. I am not sure why this happens.

Other than that should be good now. Errors are aligned with Android version and I removed (null) lines from project.pbxproj

**Comment by sche** [comment]
Thanks @JohnnyJumper 
There is still a conflict. Probably because we make quite some merges currently. Could you please fix conflicts next Tuesday (this will be a planning day for us with probably no merges), and we will merge them?

**Comment by sche** [comment]
@JohnnyJumper do you plan to resolve conflicts this week?

**Comment by JohnnyJumper** [comment]
@sche After resolving the conflicts and updating to the gnosis:main branch I start to receive a firebase error on local build. Probably some missing config causes that. 
To be specific here is the error.
Screen Shot 2021-05-12 at 4 09 21 AM

**Comment by DmitryBespalov** [comment]
@JohnnyJumper @sche I think this is ok to merge, we'll fix that thing on our end. @MouazAlzahabi it'd be good to create an improvement to not use remote config if the Firebase config files are not available (I think it is already implemented in our code for the Firebase Analytics)

**Review by sche — APPROVED** [review]

**Comment by DmitryBespalov** [comment]
This will be released in approx. 3 weeks

**Pull request merged**
